Key Binds API

EnnuiL opened the pull request:
A built-from-scratch API that covers Fabric Key Binding API's functionality and more

The PR was planned in two stages: Blue and Pink. The implemented feature set of Blue was:
- Key binds now support being dynamically disabled, making them completely invisible to the game's eyes
- The Key Binds screen now shows a tooltip on conflicting keys showing the other key binds that use the same key, and adds a `(!)` indicator to the key text itself for accessibility reasons
- The API exposes some methods meant to help with mod intercompatibility

Currently, work is being done on the Pink feature set:
- [ ] A custom config file at `./config/qsl/key_binds.json` allowing for our own key bind extensions (WIP)
- [ ] Key binds now accepts key chords, expanding the possibilities for modded key binds (WIP)
- [ ] Key binds will contain key predicates, which essentially points all the use cases where it will be used. This will allow multiple key binds that never overlaps to share the same key

Outdated but relevant information about the stages is available in #61

EnnuiL committed:
Begin Key Bindings API
Oh, this is definitely not ready yet, I just needed an easy way to sync progress between computers

EnnuiL committed:
Begin dynamic key binding system
It's alive!

EnnuiL committed:
Further progress on Key Binding API

EnnuiL committed:
More API progress
If everything goes right, tomorrow we'll have a PR!

EnnuiL committed:
All ready for a draft PR!

EnnuiL committed:
Hm, maybe it's not bad after all

TheGlitch76 commented:
Now is the time for any last-minute bikesheds on where to put the keybindings API, since I know people are unhappy with `gui`

OroArmor commented:
`gui` isn't perfect, but its the closest large group in my opinion.

you could make an `input` or `client` group, but then we end up with 30 different modules

OroArmor commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/api/KeyBindingRegistry.java:
```diff
@@ -0,0 +1,78 @@
+import org.quiltmc.qsl.key.bindings.impl.KeyBindingRegistryImpl;
+
+@Environment(EnvType.CLIENT)
+public class KeyBindingRegistry {
```
Javadocs for this class.

OroArmor commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/api/KeyBindingRegistry.java:
```diff
@@ -0,0 +1,78 @@
+		return KeyBindingRegistryImpl.getKeyBinding(translationKey);
+	}
+
+	/**
+	 * Gets the state of the key binding.
+	 *
+	 * @param key the key binding
+	 * @return {@code true} if the key binding is enabled, {@code false} otherwise
+	 */
+	public static boolean isEnabled(KeyBinding key) {
+		return KeyBindingRegistryImpl.isEnabled(key);
+	}
+
+	/**
+	 * Sets the state of the key binding.
+	 *
+	 * @param key the key binding
+	 * @param enabled the new state
+	 */
+	public static void setEnabled(KeyBinding key, boolean enabled) {
+		KeyBindingRegistryImpl.setEnabled(key, enabled);
+	}
```
What exactly does `enabled` mean? Personally I know, but its a little vague here and should be clarified to be that the `KeyBinding` does not work.

OroArmor commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/impl/KeyBindingRegistryImpl.java:
```diff
@@ -0,0 +1,133 @@
+
+		for (KeyBinding otherKey : quiltKeys.keySet()) {
+			if (key == otherKey || key.getTranslationKey() == otherKey.getTranslationKey()) {
+				LOGGER.error("Attempted to register %s, but a conflicting key has already been registered!", key.getTranslationKey());
```
Clarify what `conflicting` means in this context. I would assume the same keybind, but its the translation key.

OroArmor replied:
In addition, i would have a separate error for when the same keybinding is registered twice.

OroArmor commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/mixin/client/KeyBindingEntryMixin.java:
```diff
@@ -0,0 +1,100 @@
+
+@Environment(EnvType.CLIENT)
+@Mixin(KeyBindingEntry.class)
+public abstract class KeyBindingEntryMixin extends KeyBindingListWidget.Entry implements ConflictTooltipOwner {
```
I would add a way to show that a keybinding is disabled. I'm sure there is a texture that would work for this

EnnuiL replied:
When a keybind's disabled, not only it won't work, but it'll be hidden in the game; It's that way because it would be perfect for the usecase of disabling a keybind in-game without requiring a restart (ex: Ok Zoomer's extra keybinds).

However, thinking about it, perhaps I should split the `enabled` state into `enabled` and `visible`; I'll see what I can do, since there's still room for major changes like that

OroArmor replied:
Ah ok. I thought it would still show up, but have no effect in game. makes sense now.

EnnuiL replied:
After some thought, having a separate state for only disabling it functionally doesn't really make sense; Mods can already do that themselves

OroArmor reviewed (approved):
Looks great! I know you talked about supporting some combination keybinds, not sure if that should be an extra PR or not, but its definitely something to discuss further

EnnuiL committed:
Address suggestions

EnnuiL committed:
Minor tweaks and enhancements

EnnuiL committed:
Progress...?

BasiqueEvangelist commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/api/KeyBindingRegistry.java:
```diff
@@ -0,0 +1,117 @@
+	 * Sets the state of the key binding.
+	 *
+	 * <p>This method allows for disabling the key binding, making it invisible
+	 * to the game. options.txt is the only expection, who still stores the disabled
```
```suggestion
	 * to the game. options.txt is the only exception, who still stores the disabled
```

EnnuiL replied:
I am _sooooo_ good at speaking languages

BasiqueEvangelist commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/api/KeyBindingRegistry.java:
```diff
@@ -0,0 +1,117 @@
+	 * <p>The state of a key binding determines whenever it's disabled or not.
+	 *
+	 * <p>When a keybind is disabled, it is effectively hidden from the game,
+	 * being non-existent to it. options.txt is the only expection, who still
```
```suggestion
	 * being non-existent to it. options.txt is the only exception, who still
```

BasiqueEvangelist commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/impl/KeyBindingRegistryImpl.java:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	public static void setEnabled(KeyBinding key, boolean enabled) {
+		if (quiltKeys.containsKey(key)) {
```
This should throw an exception if the key isn't a quilt key

EnnuiL replied:
Good idea

EnnuiL replied:
And done; It's currently the same exception for both non-registered keys and vanilla ones, but I'll eventually split it

BasiqueEvangelist commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/impl/KeyBindingRegistryImpl.java:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	public static void registerKeyBindingManager(GameOptions options, KeyBinding[] allKeys) {
+		keyBindingManagers.add(new KeyBindingManager(options, allKeys));
```
Is there any case where there is *more than one* key binding manager?

EnnuiL replied:
Well, if another mod creates another GameOptions instance, then yes, there would be more than one manager, however, I'll probably ditch this system; It won't really be future-proof for a possible creation of a separate keybind options file

BasiqueEvangelist replied:
> if another mod creates another GameOptions instance

Do mods ever create GameOptions instances, though? And if they do, why?

EnnuiL replied:
I'm not aware of any mods that do that, but well, mods do cursed stuff all the time, so I thought that having more than one GameOptions instance would be plausible

BasiqueEvangelist commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/impl/KeyBindingRegistryImpl.java:
```diff
@@ -0,0 +1,134 @@
+
+	public static KeyBinding registerKeyBinding(KeyBinding key, boolean enabled) {
+		if (key == null) {
+			LOGGER.error("Attempted to register a null key bind!");
```
Why is this just logged and not an immediate exception?

EnnuiL replied:
Like, throw an NPE? I already return it immediately after that message

BasiqueEvangelist replied:
You just return `null`, not an NPE.
This would mean that an NPE is thrown the first time the mod tries to use the keybind, which is... not optimal.

EnnuiL replied:
I already return `null` one line after the log message though, I don't let the rest of the code run and make bad things happen

EnnuiL replied:
Oh boy, that was a confusing discussion, but the idea (use exceptions instead of returning null) has been implemented

EnnuiL committed:
Fix typos and checkstyle issues

EnnuiL committed:
Apply suggestions, fix AW

EnnuiL committed:
Small progress!

EnnuiL committed:
But wait, there's more!

EnnuiL committed:
Improve tooltip formatting

EnnuiL committed:
Progress progress progress

EnnuiL committed:
Fix key binding manager

EnnuiL committed:
Add "Middle Button to Unbind" behavior

EnnuiL committed:
oof

EnnuiL committed:
Fix conflict tooltips

EnnuiL committed:
Add registry test, progress

EnnuiL committed:
Mappings update

EnnuiL committed:
Finish the v1 feature set

EnnuiL committed:
Fix issues

EnnuiL committed:
whyyyyyyyyy

EnnuiL committed:
Remove "Middle Click to Unbind"
Because Vanilla actually has a way to unbind keys: "Select and Press Esc to Unbind"

OroArmor commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/api/KeyBindingRegistry.java:
```diff
@@ -0,0 +1,135 @@
+	 *
+	 * <pre>
+	 * {@code
+	 * public static final KeyBinding EXAMPLE_KEY_BIND = KeyBinding.registerKeyBinding(
```
```suggestion
	 * public static final KeyBinding EXAMPLE_KEY_BIND = KeyBindingRegistry.registerKeyBinding(
```

EnnuiL replied:
Oh whoops, I'll fix that

OroArmor commented on library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/impl/ConflictTooltipOwner.java:
```diff
@@ -0,0 +1,29 @@
+import net.minecraft.text.Text;
+
+@Environment(EnvType.CLIENT)
+public interface ConflictTooltipOwner {
```
This could either be exposed as API as it could have some functionality or make it slightly clearer that it's internal with `@ApiStatus.Internal`

EnnuiL replied:
Ah, I was thinking of replacing the interface mixin with the two accessors that I've done for resource pack tooltips; I think it's much cleaner than the current one

EnnuiL replied:
...and nevermind, while the accessor does simply a bit of code, ConflictTooltipOwner will still be needed;
I'll mark it as internal

EnnuiL committed:
Begin refactor to Key Binds, cleanup tooltip mixin
More will come after the mappings update

EnnuiL committed:
Attempt to remove an unchangable file
how did this happen

EnnuiL committed:
Add back Brazilian Portuguese translation (1)
and despite the previous commit, git still hates this file
why

EnnuiL committed:
Update to 22w06a, fix license and testmod

EnnuiL committed:
Depend on resource loader
Required for translatable texts to work on the tooltips

EnnuiL committed:
Use the mappings refactor (I)
I'll finish it at my other PC

LambdAurora commented on library/gui/key_binds/src/main/java/org/quiltmc/qsl/key/binds/api/KeyBindRegistry.java:
```diff
@@ -0,0 +1,135 @@
+	 * @param includeVanilla {@code true} if vanilla entries should be included, else {@code false}
+	 * @return a map containing all modded (and optionally vanilla) key binds
+	 */
+	public static Map<KeyBind, Boolean> getAllKeyBinds(boolean includeVanilla) {
```
This really should use fastutil's Object2BooleanMap.

EnnuiL replied:
I'll probably abandon the Map<KeyBind, Boolean> approach once on the Pink stage, because having some sort of `QuiltKeyBinds` would probably be better than a map between keybinds and the extra info

i509VCB replied:
Probably would be a good idea to encapsulate the inner map so the api can be a little more resistant to change.

EnnuiL committed:
Use the mappings refactor (II)

EnnuiL committed:
Actually fix the Brazilian Portuguese translation
Viva!

EnnuiL committed:
Move to QSL event entrypoints

EnnuiL committed:
Fix conflict tooltips

EnnuiL committed:
Begin Pink's config system and key chords

LambdAurora commented:
> A custom config file at `./config/qsl/key_binds.json` allowing for our own key bind extensions (WIP)

Now that I'm thinking of it, probably should be `./config/quilt/key_binds.json`.

EnnuiL committed:
Apply license; Use sorted maps for chords

EnnuiL committed:
Add chord support for conflict warnings (I)

EnnuiL commented:
> > A custom config file at `./config/qsl/key_binds.json` allowing for our own key bind extensions (WIP)
> 
> Now that I'm thinking of it, probably should be `./config/quilt/key_binds.json`.

I thought about it (after all, FAPI used `./config/fabric/`), however, I don't know; I'd think `./config/quilt/` would instead be used by something like the Quilt Loader, and not QSL

EnnuiL committed:
Uh, licença

EnnuiL committed:
Fix conflict tooltips

EnnuiL committed:
Begin to use FastUtils, add overflow hiding, more!

EnnuiL committed:
Pacify Checkstyle
wait, so, tabs are actually okay?

EnnuiL committed:
Fix lack of conflicts with the chord map

EnnuiL committed:
Allow for setting default chords

EnnuiL committed:
Quiltify Key Binds API, tackle down config issues

EnnuiL committed:
Push changes done to config
I have a plan to properly fix it, but now, it's icon time

EnnuiL committed:
Fix stuff
no thoughts, head empty

EnnuiL committed:
Begin to overhaul toggling
Also, begin to implement unused_key_binds

EnnuiL committed:
Recalculating
mind's currently empty, but i'll find a better way to do things

EnnuiL committed:
Refactor the key bind registry
It's gone!

EnnuiL committed:
Begin port to 1.19

EnnuiL committed:
Apply licenses

EnnuiL committed:
Fix config system

EnnuiL committed:
Reignite work on Key Binds API

EnnuiL committed:
keyboard mash

EnnuiL committed:
Partially port key binds api to 1.20

EnnuiL committed:
Merge branch '1.20' into 1.18
